### ISSUE TITLE
HttpContext.Response.Body writes an empty NewLine when no bytes were written

### ISSUE THREAD

**Issue by ststeiger** [dotnet/aspnetcore#37922]
### Description

System.IO.StreamWriter writes an empty \r\n on dispose. (Flush?)
It should not write anything. 

Problem occurs when it writes it to HttpContext.Response.Body. 
Then the response counts as started, and I can't write a status code afterwards, because the runtime won't let me (response already started). 
```
// strm = HttpContext.Response.Body
using (System.IO.TextWriter output = new System.IO.StreamWriter(strm, new System.Text.UTF8Encoding(false), 4096, true))
{
    throw new System.Exception("boom");
    // I do things only after here - but the error also occurs if the exception is before
   // if the using is omitted, i don't get the "response already started"-message
} // End Using output 
```

### Reproduction Steps

Have an SQL-Handler for an URL: e.g.
https://localhost:44314/ajax/AnySelect.ashx?sql=GetChecklistData.sql&format=1&__cl_uid=EB159A9C-E69F-49F4-B10E-3A4825973E46

Which is handled by SQL-Middleware (getting parameters from url)
https://github.com/ststeiger/TestPWA/blob/4baa1ac47102a762d85f1efe96d4a7a81412a11f/TestPWA/Code/DB/SqlMiddleware.cs#L89
(problematic place at line 89)

Which reads the GetChecklistData.sql from disk and executes it
https://github.com/ststeiger/TestPWA/blob/4baa1ac47102a762d85f1efe96d4a7a81412a11f/TestPWA/Code/SQL/GetChecklistData.sql

If the statement 
-- DECLARE @__cl_uid uniqueidentifier; 
-- SET @__cl_uid = 'EB159A9C-E69F-49F4-B10E-3A4825973E46'; 
is **NOT** commented out, the SQL-produces an error (parameter already declared). 
So far so good - that's not the problem. 

But the handling of the error afterwards. 

### Expected behavior

I can write an error-response, because **nothing has been written into HttpContext.Response "so far"** ...

I write data to the HttpResponse in 
public static async System.Threading.Tasks.Task<System.Exception> AsJSON(
of file 
https://github.com/ststeiger/TestPWA/blob/4baa1ac47102a762d85f1efe96d4a7a81412a11f/Dapper/__AsJSON.cs#L292

I expect that I can write to the HttpResponse with new System.IO.StreamWriter with LeaveOpen on true. 
If nothing gets written in AsJSON, I expect to be able to write a statuscode != 200 + ErrorMessage afterwards. 

### Actual behavior

I **cannot** write an error-response, because streamWriter has already written a **newLine** into HttpContext.Response, so the response counts as **"already begun"**. 

In File
https://github.com/ststeiger/TestPWA/blob/4baa1ac47102a762d85f1efe96d4a7a81412a11f/Dapper/__AsJSON.cs#L292
I expect that I can write to the HttpResponse with new System.IO.StreamWriter with LeaveOpen on true. 
If nothing gets written in AsJSON, I expect to be able to write a statuscode != 200 + ErrorMessage afterwards. 
But I have to create a derived-class "AAAA" that does nothing on dispose (if there is an error in the sql-statement), if I want to be able to write anything to the response afterwards. 

Why does it do this ? 
The error is not that something gets written into the stream in AsJSON. 
I can comment it out, and it still happens, if I just throw an error instead of executing await AsJson. 
Everything works fine if there is no statement 
```
using (System.IO.TextWriter output = new System.IO.StreamWriter(strm, new System.Text.UTF8Encoding(false), 4096, true))
{
 throw new Exception();
}
```
so there also is nothing written into the response before. 
Where does this NewLine (or whatever) come from ?

### Regression?

Not to my knowledge. 

### Known Workarounds

Create a class "AAAA" derived from StreamWriter that doesn't execute `base.dispose` on Exception. 
```
  public class AAAA
        : System.IO.StreamWriter
```
https://github.com/ststeiger/TestPWA/blob/4baa1ac47102a762d85f1efe96d4a7a81412a11f/Dapper/__AsJSON.cs#L303

But that probably leads to a memory-leak (StreamWriter not disposed properly). 

### Configuration

.NET 5
Windows 10 x64, all updates and servicepacks applied. 
Using IIS Express.
To my knowledge, it's not specific to this platform. 

### Other information

I think on StreamWriter.Dispose, there is something writing a NewLine/WhiteSpace, e.g. Flush. 
This probably causes the response to be counted as *Response has begun".

**Comment by Tratcher** [comment]
I pulled the latest copy of the repro from https://github.com/ststeiger/SimpleStreamWriterBugReproduction and enabled kestrel connection logging. Here's what I see:

```
info: Microsoft.AspNetCore.Hosting.Diagnostics[1]
      Request starting HTTP/1.1 GET http://localhost:5000/ajax/WithStreamWriter.ashx - -
trce: Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware[2]
      All hosts are allowed.
dbug: Microsoft.AspNetCore.Routing.Matching.DfaMatcher[1000]
      No candidates found for the request path '/ajax/WithStreamWriter.ashx'
dbug: Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware[2]
      Request did not match any endpoints
dbug: Microsoft.AspNetCore.Server.Kestrel.Core.Internal.LoggingConnectionMiddleware[0]
      WriteAsync[101]
      48 54 54 50 2F 31 2E 31  20 32 30 30 20 4F 4B 0D   HTTP/1.1  200 OK.
      0A 44 61 74 65 3A 20 46  72 69 2C 20 32 39 20 4F   .Date: F ri, 29 O
      63 74 20 32 30 32 31 20  31 37 3A 35 32 3A 30 37   ct 2021  17:52:07
      20 47 4D 54 0D 0A 53 65  72 76 65 72 3A 20 4B 65    GMT..Se rver: Ke
      73 74 72 65 6C 0D 0A 54  72 61 6E 73 66 65 72 2D   strel..T ransfer-
      45 6E 63 6F 64 69 6E 67  3A 20 63 68 75 6E 6B 65   Encoding : chunke
      64 0D 0A 0D 0A                                     d....
```
This shows the response being flushed as expected when the an exception is thrown and the `using` calls `Dispose` on the TextWriter. This `0D 0A 0D 0A` is the standard `\r\n\r\n` used to mark the end of HTTP headers, it's not part of the body.
```
StatusCode cannot be set because the response has already started.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ThrowResponseAlreadyStartedException(String value)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.set_StatusCode(Int32 value)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.Microsoft.AspNetCore.Http.Features.IHttpResponseFeature.set_StatusCode(Int32 value)
   at Microsoft.AspNetCore.Http.DefaultHttpResponse.set_StatusCode(Int32 value)
   at SimpleStreamWriterBugReproduction.Startup.<>c.<<Configure>b__5_7>d.MoveNext() in C:\temp\SimpleStreamWriterBugReproduction\SimpleStreamWriterBugReproduction\Startup.cs:line 106
```
That's the expected failure when trying to change the status code after the headers have been flushed.
```
dbug: Microsoft.AspNetCore.Server.Kestrel.Core.Internal.LoggingConnectionMiddleware[0]
      WriteAsync[5]
      30 0D 0A 0D 0A                                     0....
dbug: Microsoft.AspNetCore.Server.Kestrel[9]
      Connection id "0HMCR0J9ELM5C" completed keep alive response.
info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished HTTP/1.1 GET http://localhost:5000/ajax/WithStreamWriter.ashx - - - 200 - - 1751.5690ms
```
This is kestrel closing out the response. `30 0D 0A 0D 0A` is `0\r\n\r\n` which is the standard chunked end-of-body sequence, it's also not considered part of the body.

When I replaced `context.Response.Body` with a MemoryStream in the repro then no data was written to the memory stream.

I don't see any signs of an extra `\r\n` written in these scenarios.

The AAA.ashx endpoint with your derived StreamWriter worked fine because it was set to skip the call to Flush on Dispose.

Here's a different pattern you can use to avoid this Flush/Exception ordering issue. Invert the using and the try block, and also be sure to check for HasStarted and Abort the response (or rethrow the exception) so the client doesn't think the response is valid.
```C#
                    appBuilder.Use(
                        async delegate (Microsoft.AspNetCore.Http.HttpContext context, Func<Task> next)
                        {
                            using (System.IO.TextWriter output = new System.IO.StreamWriter(context.Response.Body, new System.Text.UTF8Encoding(false), 4096, true))
                            // using (System.IO.StreamWriter output = new System.IO.StreamWriter(context.Response.Body, new System.Text.UTF8Encoding(false), 4096, true) { AutoFlush = false })
                            // using (AAAA output = new AAAA(context.Response.Body, new System.Text.UTF8Encoding(false), 4096, true) { ExceptionFetcher = delegate { return caughtByCodeAndReturnedFromFunction; } })
                            {
                                try
                                {
                                    throw new Exception("SQL-Execution Error");
                                } // End Try 
                                catch (Exception ex)
                                {
                                    if (context.Response.HasStarted)
                                    {
                                        context.Abort();
                                        return;
                                    }
                                    try
                                    {
                                        context.Response.StatusCode = (int)System.Net.HttpStatusCode.InternalServerError;
                                        // context.Response.Headers["X-Error-Message"] = "Incorrect username or password";
                                        context.Response.Headers["X-Error-Message"] = ex.Message;

                                        // context.Response.ContentType = "text/plain";
                                        // byte[] msg = System.Text.Encoding.UTF8.GetBytes("this is a test");
                                        // await context.Response.Body.WriteAsync(msg, 0, msg.Length);

                                        context.Response.ContentType = "application/json";
                                        await Helpers.ToJSON(context.Response.Body, ex);
                                    }
                                    catch (Exception ex2)
                                    {
                                        Console.WriteLine(ex2.Message);
                                        Console.WriteLine(ex2.StackTrace);
                                        NativeMethods.ShowMessageBox(ex2.Message, ex2.StackTrace);
                                    }
                                } // End Catch
                            }
                        }
```

**Comment by ststeiger** [comment]
But of what use is `context.Abort();` to me ? 
If I wanted to do that, I could just as well let the request fail after re-throwing the exception after logging it in the catch. 
What I want to do is to pass to the application that an error occured, with further information on the exact nature of the error. 
A HTTP 500 is not very informative. 
Not to mention that I can't write a HTTP 500 either - the browser has to notice this on its own. 

What I'm trying to achive is something along the lines of: 

```
{
   status: "OK"
   data: { live-serialized JSON that's directly written from the DataReader with SerialAccess to the response stream - no intermediate in-memory list - amount of data might be very large }
}
```

and if there's an error in the sql-statement, I want this to happen: 

```
{
   status: "ERROR"
   exception: { SerializedDotNetException } /* ok, it can be argued if outputting exception details in production is a good idea - but for development, it would be nice */
}
```
So that I can do this on the client-side: 
```
let answer = await fetch("AAA.ashx").json();
if(answer.status !== "OK") {alert(answer.error);  return; }...
```
where alert is substituted with something like jquery-confirm. 

As you have already noticed, if it flushes an empty streamwriter's non-content into its base-http-response-stream, the response has started, and I can't do anything anymore. 
But the question is - why does it flush - if it has NOTHING to flush ???
That makes no sense to me. 

Also, if you assume that it's correct behaviour to flush "empty" from a streamwriter into a response - which triggers all the http-headers that are sent: 
 i don't know if there are any instances in which that makes sense at all - but there might be - then I should be able to override this behaviour by just creating a class that inherits from streamwriter, and then overriding the flush methods. 
I tried passing the exception to this derived class, and not flush if there is an exception. That doesn't work because Dispose in streamwriter doesn't call the virtual flush methods, but its own private one - so overriding the flush methods has no effect (on dispose). 

The only instance in which 
```
if (context.Response.HasStarted)
{
    context.Abort();
    return;
}
```
should be relevant, is, when an error happens during JSON-serialization after a few rows have already been successfully written to the output stream. e.g. when System.Text.JSON doesn't handle the serialization of a sudden System.DbNull.Value, e.g. when you're serializing a System.Data.DataTable instead of SQL (which doesn't happen in this example since it only serializes from DataReader), or on an arithmetic overflow, or on a Div0-exception, or any other Runtime-and-not-CompileTime-errors with "row/field-scope". 

And when I have to rip the sourcecode of streamwriter out of the .NET framework - to modify the flush on dispose - I have the added complication that the resulting code can only be .NET 5.0, and not netstandard 2.0, because it doesn't compile on netstandard 2.0.

**Comment by ststeiger** [comment]
One more thing: 
There is even more wrong in all of this: 

If I call a sql-script like `SELECT * FROM bla WHERE CL_UID =  @__cl_uid` 

And pass __cl_uid = "null";

then on `while (await dr.ReadAsync())` 
instead of "Error converting string to 'uniqueidentifier'. " 
i get ` {"Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead."}`

When using Newtonsoft.JSON and StreamWriter, this is fixed by using 
`FlushAsyncInternal(true, true, _charBuffer, _charPos).Wait();`
in StreamWriter.Dispose instead of `Flush(flushStream: true, flushEncoder: true);`. 

But fun fact is, although System.Text.Json uses Utf8JsonWriter,  the same thing happens. 
So you get the wrong error message (again, before writing anything to stream). 
Seems like Utf8JsonWriter also does a synchronous flush on dispose, which is disallowed when sync-mode is turned off. 
Nastly little bugs.  
Have already been wondering before why it didn't hit this one... 

[![127 little bugs][1]][1]


```
            // If using Kestrel:
            services.Configure<Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions>(options =>
            {
                options.AllowSynchronousIO = false;
            });

            // If using IIS:
            services.Configure<IISServerOptions>(options =>
            {
                options.AllowSynchronousIO = false;
            });
```

  [1]: https://i.stack.imgur.com/CKqcH.jpg

**Comment by Tratcher** [comment]
> But the question is - why does it flush - if it has NOTHING to flush ???
> That makes no sense to me.

I won't pretend to know why, but that's the behavior and there's no opportunity to change it until at least 7.0, if ever. It sounds like you've avoided this by moving to System.Text.Json.

> Seems like Utf8JsonWriter also does a synchronous flush on dispose, which is disallowed when sync-mode is turned off.
Nasty little bugs.

Utf8JsonWriter does implement IAsyncDisposable, so you should be able to avoid this one with an async using.

On a different note, we actually found using statements to be problematic for serializes for exactly this reason. If an exception happens we don't want to flush the content, we want to change or abort the response. We moved to an explicit dispose pattern in many of the MVC serializers.

**Comment by ststeiger** [comment]
Didn't know async-using existed already. 
Yes that solved the problem. 

> I won't pretend to know why, but that's the behavior and there's no opportunity to change it until at least 7.0, if ever. It sounds like you've avoided this by moving to System.Text.Json.
I understand - I would not change that either with one week to LTS-release; too many things that could go wrong.

On a different note about the serializers as well: 
The serializer misses 
`System.Text.Json.JsonSerializer.SerializeAsync(writer, obj, options);`
it only has 
`await System.Text.Json.JsonSerializer.SerializeAsync(stream, obj, thisError.Data.GetType(), options);`

I'll file a new issue for that. 

Example of the problem:

```
  public class SessionExpiredException
        : System.Exception
    { }



    [System.Serializable()]
    public class AJAXException
        // : System.Exception
    {


        protected System.Exception m_Exception = null;
        protected string m_Message = null;


        public bool sessionExpired
        {
            get
            {
                if (m_Exception != null)
                {
                    if (m_Exception.GetType() == typeof(SessionExpiredException))
                        return true;
                }

                return false;
            }
        } // sessionExpired


        public string message
        {
            get
            {
                if (sessionExpired)
                    return "Session expired";

                if (!string.IsNullOrEmpty(m_Message))
                    return m_Message;

                if (m_Exception != null)
                    return m_Exception.Message;

                return "";
            }

            set
            {
                m_Message = value;
            }
        } // message


        public string originalMessage
        {
            get
            {
                if (m_Exception != null)
                {
                    if (!string.IsNullOrEmpty(m_Exception.Message))
                        return m_Exception.Message;
                }

                return "";
            }
        } // originalMessage


        public string source
        {
            get
            {
                if (m_Exception != null)
                {
                    if (!string.IsNullOrEmpty(m_Exception.Source))
                        return m_Exception.Source;
                }

                return "";
            }
        } // source


        public string stackTrace
        {
            get
            {
                if (m_Exception != null)
                {
                    if (!string.IsNullOrEmpty(m_Exception.StackTrace))
                        return m_Exception.StackTrace;
                }

                return "";
            }
        } // stackTrace


        public System.Exception innerException
        {
            get
            {
                return m_Exception;
            }
        } // innerException


        public AJAXException(System.Exception exBaseException)
        {
            // MyBase.New(exBaseException.Message, exBaseException)
            // Me.m_RealStackTrace = New StackTrace(exBaseException)
            this.m_Exception = exBaseException;
        } // Constructor


        public AJAXException(string strMessage, System.Exception exBaseException)
        {
            this.message = strMessage;
            this.m_Exception = exBaseException;
        } // Constructor




        public async System.Threading.Tasks.Task ToJSON(System.IO.Stream s)
        {
            System.Text.Json.JsonSerializerOptions options =
                new System.Text.Json.JsonSerializerOptions()
                {
                    Converters = {
                        new TestPWA.ExceptionConverter()
                    // ,new TestPWA.SystemTypeConverter() 
                    }
                    ,IncludeFields = true
                    ,WriteIndented = true
                }
            ;


            System.Text.Json.JsonWriterOptions writerOptions = new System.Text.Json.JsonWriterOptions();
            // if (format.HasFlag(RenderType_t.Indented))
            writerOptions.Indented = true;

            System.Exception thisError = this.innerException;


            using (System.Text.Json.Utf8JsonWriter writer = new System.Text.Json.Utf8JsonWriter(s, writerOptions))
            {
                writer.WriteStartObject();

                writer.WritePropertyName("hasError");
                writer.WriteBooleanValue(true);

                writer.WritePropertyName("error");



                writer.WriteStartObject();


                writer.WritePropertyName("message");
                writer.WriteStringValue(this.message);

                writer.WritePropertyName("originalMessage");
                writer.WriteStringValue(this.originalMessage);


                writer.WritePropertyName("sessionExpired");
                writer.WriteBooleanValue(this.sessionExpired);

                writer.WritePropertyName("source");
                writer.WriteStringValue(this.source);

                writer.WritePropertyName("stackTrace");
                writer.WriteStringValue(this.stackTrace);



                writer.WritePropertyName("innerException");


                int innerExceptionCount = 0;

                writer.WriteStartObject();
                while (thisError != null)
                {
                    writer.WritePropertyName("Message");
                    writer.WriteStringValue(thisError.Message);

                    writer.WritePropertyName("StackTrace");
                    writer.WriteStringValue(thisError.StackTrace);

                    writer.WritePropertyName("Source");
                    writer.WriteStringValue(thisError.Source);

                    writer.WritePropertyName("HelpLink");
                    writer.WriteStringValue(thisError.HelpLink);

                    // Do NOT serialize this.
                    // writer.WritePropertyName("TargetSite");
                    // writer.WriteStringValue(thisError.TargetSite.Name);

                    writer.WritePropertyName("Type");
                    writer.WriteStringValue(thisError.GetType().FullName);

                    // writer.WritePropertyName("Data");
                    // AAAAAAAAAAAaaaaaaaaaaaaaaaaaaaaaaaaaaaah !
                    // System.Text.Json.JsonSerializer.Serialize(writer, thisError.Data, options);


                    // AAAAAAAAAAAaaaaaaaaaaaaaaaaaaaaaaaaaaaah !
                    // await writer.FlushAsync();
                    // await System.Text.Json.JsonSerializer.SerializeAsync(s, thisError.Data, options);
                    // await s.FlushAsync();

                    //    await System.Text.Json.JsonSerializer.SerializeAsync(s, thisError.Data, thisError.Data.GetType(), options);
                    // System.Text.Json.JsonSerializer.SerializeAsync(writer, thisError.Data, options);

                    if (thisError.InnerException != null)
                    {
                        innerExceptionCount++;
                        System.Console.WriteLine(thisError.InnerException);
                        writer.WritePropertyName("InnerException");
                        writer.WriteStartObject();
                    } // End if (thisError.InnerException != null) 

                    thisError = thisError.InnerException;
                } // Whend 

                for (int i = 0; i < innerExceptionCount; ++i)
                {
                    writer.WriteEndObject(); // InnerExceptions
                } // Next i 

                writer.WriteEndObject(); // innerException
                writer.WriteEndObject(); // Error
                writer.WriteEndObject(); // end object  

                await writer.FlushAsync();
            } // End Using writer 

        } // End Task ToJSON 

        
    } // AJAXException
```

**Comment by ststeiger** [comment]
Added to issue https://github.com/dotnet/runtime/issues/61221
